prometheus_hardware_exporter/collector.py

```diff
@@ -853,7 +856,10 @@ class RedfishCollector(BlockingCollector):
     def __init__(self, config: Config) -> None:
         """Initialize RedfishHelper instance."""
         super().__init__(config)
-        self.redfish_helper = RedfishHelper(config)
+        self.config = config
+        self.discover_redfish_services = get_cached_discover_function(
```

Comment:
**issue (blocking):** 

I don't get why we want to move discover to here. This somehow break the boundary of RedfishHelper and Collector.

The only path for Collector to get information from Redfish is through RedfishHelper.

Reply:
I moved the discover method outside the RedfishHelper class because the caching was not working while using the context manager. Since the object of the class is instantiated newly each time we enter the context manager and then destroyed at the end, the caching doesn't work and the delay for discovering the redfish services is always present.

Reply:
Made some changes to both the context manager and the discover method.

- The `__enter__` method in the context manager does not try to handle the exception which may be caused while getting the new redfish session. If an error occurs, the exception is raised and the caller will need to handle it.
In our case, this exception is captured and handled by the try...except statements wrapping the context manager in the RedfishCollector.fetch method.

- The discover method is now a static method inside the RedfishHelper class. This is used by the RedfishCollector class to discover redfish services and we can keep all the functionality within the RedfishHelper class itself. I've tested the cache and it works as well.

Reply:
Create redfish discover service here still not a good practice.
But the solution for this maybe separate the ContextManager and RedfishHelper.

I will put that in refactoring task instead of here.